### ISSUE TITLE
WASMFS disables "emulated posix tcp Sockets over WebSockets"

### ISSUE THREAD

**Issue by manin** [emscripten-core/emscripten#22907]
**Version of emscripten/emsdk:**

> emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.70 (b53978ee3f540dc74761eba127aa7f1b8761a125)
> clang version 20.0.0git (https:/github.com/llvm/llvm-project f52b89561f2d929c0c6f37fd818229fbcad3b26c)
> Target: wasm32-unknown-emscripten
> Thread model: posix

Building with WASMFS disables [emulated posix tcp Sockets over WebSockets](https://emscripten.org/docs/porting/networking.html#emulated-posix-tcp-sockets-over-websockets). We’ve ported a large codebase to Emscripten that relies on this socket emulation, and it works well. However, we also want to use OPFS, but currently, WASMFS and POSIX socket emulation cannot be used together.

Specifically:

* Every socket syscall becomes a stub in [system/lib/wasmfs/syscalls.cpp](https://github.com/emscripten-core/emscripten/blob/e6ed9d3fc2821a72cc8ded4a02cac72b9f306f88/system/lib/wasmfs/syscalls.cpp#L1697)

* Additionally, emscripten_set_socket_*_callback symbols are not exported, resulting in link-time errors. For example, we use emscripten_set_socket_message_callback in a reactor, and this fails to link.

Is there any workaround to use OPFS and posix socket emulation? Alternatively, are there any plans to support this combination?

The [emulated posix tcp Sockets over WebSockets](https://emscripten.org/docs/porting/networking.html#emulated-posix-tcp-sockets-over-websockets) documentation should also be updated to reflect that it won’t work with -sWASMFS.

**Comment by kripken** [comment]
The SOCKFS backend for the old FS needs to be ported to WasmFS. Or, there was an idea to write a general shim that would run any of the old backends in WasmFS. I am not aware of current work on either, but this is definitely something that would be good to have.

@sbc100 I recall you were doing some cleanups on sockets code recently - were you aiming to do this perhaps?

**Comment by kripken** [comment]
Yeah, I think a compat layer is worth looking into. It seems like it could get us SOCKFS, IDBFS, WORKERFS etc all in one go. That could get many (most?) of the remaining tests passing on WasmFS.

(I feel this is important, but there may be more urgent things overall.)
